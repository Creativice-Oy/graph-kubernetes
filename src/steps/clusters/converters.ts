import { createIntegrationEntity } from '@jupiterone/integration-sdk-core';
import { Cluster } from '@kubernetes/client-node';
import { Entities } from '../constants';

function getClusterKey(id: string) {
  return `kube_cluster:${id}`;
}

export function createClusterEntity(data: Cluster) {
  const { caData, caFile, ...clusterData } = data;

  return createIntegrationEntity({
    entityData: {
      source: {
        ...clusterData,
      },
      assign: {
        _class: Entities.CLUSTER._class,
        _type: Entities.CLUSTER._type,
        _key: getClusterKey(data.name),
        name: data.name,
        displayName: data.name,
        server: data.server,
        skipTlsVerify: data.skipTLSVerify,
      },
    },
  });
}
