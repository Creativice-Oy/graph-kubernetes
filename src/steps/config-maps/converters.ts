import * as k8s from '@kubernetes/client-node';
import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { Entities } from '../constants';

export function createConfigMapEntity(configMap: k8s.V1ConfigMap) {
  /*
    ConfigMap contains 2 fields - data and binary data.
    They expose the actual data that's stored inside the map.
    We probably don't want to store this in _rawData
      1) Could be big (?)
      2) User may store something sensitive inside (even though that'd be against ConfigMap recommended use)
  */
  const { binaryData, data, ...withoutData } = configMap;

  return createIntegrationEntity({
    entityData: {
      source: withoutData,
      assign: {
        _class: Entities.CONFIGMAP._class,
        _type: Entities.CONFIGMAP._type,
        _key: withoutData.metadata?.uid,
        name: withoutData.metadata?.name,
        displayName: withoutData.metadata?.name,
        generation: withoutData.metadata?.generation,
        deletionGracePeriodSeconds:
          withoutData.metadata?.deletionGracePeriodSeconds,
        resourceVersion: withoutData.metadata?.resourceVersion,
        createdOn: parseTimePropertyValue(
          withoutData.metadata?.creationTimestamp,
        ),
        deletedOn: parseTimePropertyValue(
          withoutData.metadata?.deletionTimestamp,
        ),
        // spec fields
        immutable: withoutData.immutable,
        classification: null,
        encrypted: false,
      },
    },
  });
}
