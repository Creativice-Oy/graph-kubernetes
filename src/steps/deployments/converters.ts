import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import * as k8s from '@kubernetes/client-node';
import { Entities } from '../constants';

function convertToString(value?: object) {
  if (!value) {
    return undefined;
  }

  if (typeof value !== 'string') {
    return JSON.stringify(value);
  }

  return value;
}

export function getVolumeKey(deploymentId: string, volumeName: string) {
  return `${deploymentId}/${volumeName}`;
}

export function createVolumeEntity(deploymentUid: string, data: k8s.V1Volume) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.VOLUME._class,
        _type: Entities.VOLUME._type,
        _key: getVolumeKey(deploymentUid, data.name),
        name: data.name,
        displayName: data.name,
        'ephemeral.readOnly': data.ephemeral?.readOnly,
        'hostPath.path': data.hostPath?.path,
        'hostPath.type': data.hostPath?.type,
        classification: null,
        encrypted: false,
      },
    },
  });
}

export function getContainerSpecKey(deploymentId: string, specName: string) {
  return `${deploymentId}/${specName}`;
}

export function createContainerSpecEntity(
  deploymentUid: string,
  data: k8s.V1Container,
) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.CONTAINER_SPEC._class,
        _type: Entities.CONTAINER_SPEC._type,
        _key: getContainerSpecKey(deploymentUid, data.name),
        name: data.name,
        displayName: data.name,
        image: data.image,
        imagePullPolicy: data.imagePullPolicy,
        'securityContext.allowPrivilegeEscalation':
          data.securityContext?.allowPrivilegeEscalation,
        'securityContext.privileged': data.securityContext?.privileged,
        'securityContext.procMount': data.securityContext?.procMount,
        'securityContext.readOnlyRootFilesystem':
          data.securityContext?.readOnlyRootFilesystem,
        'securityContext.runAsGroup': data.securityContext?.runAsGroup,
        'securityContext.runAsNonRoot': data.securityContext?.runAsNonRoot,
        'securityContext.runAsUser': data.securityContext?.runAsUser,
      },
    },
  });
}

export function createDeploymentEntity(data: k8s.V1Deployment) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.DEPLOYMENT._class,
        _type: Entities.DEPLOYMENT._type,
        // metadata properties
        _key: data.metadata?.uid,
        name: data.metadata?.name,
        displayName: data.metadata?.name,
        generation: data.metadata?.generation,
        deletionGracePeriodSeconds: data.metadata?.deletionGracePeriodSeconds,
        resourceVersion: data.metadata?.resourceVersion,
        createdOn: parseTimePropertyValue(data.metadata?.creationTimestamp),
        // spec properties
        minReadySeconds: data.spec?.minReadySeconds,
        paused: data.spec?.paused,
        progressDeadlineSeconds: data.spec?.progressDeadlineSeconds,
        replicas: data.spec?.replicas,
        revisionHistoryLimit: data.spec?.revisionHistoryLimit,
        'strategy.type': data.spec?.strategy?.type,
        'strategy.maxSurge': convertToString(
          data.spec?.strategy?.rollingUpdate?.maxSurge,
        ),
        'strategy.maxUnavailable': convertToString(
          data.spec?.strategy?.rollingUpdate?.maxUnavailable,
        ),
        // TODO: something about spec.template sections maybe?
        'status.availableReplicas': data.status?.availableReplicas,
        'status.collisionCount': data.status?.collisionCount,
        'status.observedGeneration': data.status?.observedGeneration,
        'status.readyReplicas': data.status?.readyReplicas,
        'status.replicas': data.status?.replicas,
        'status.unavailableReplicas': data.status?.unavailableReplicas,
        'status.updatedReplicas': data.status?.updatedReplicas,
      },
    },
  });
}
