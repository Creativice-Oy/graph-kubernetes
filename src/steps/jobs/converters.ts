import * as k8s from '@kubernetes/client-node';
import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { Entities } from '../constants';

export function createJobEntity(data: k8s.V1Job) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.JOB._class,
        _type: Entities.JOB._type,
        _key: data.metadata!.uid!,
        name: data.metadata?.name,
        displayName: data.metadata?.name,
        generation: data.metadata?.generation,
        deletionGracePeriodSeconds: data.metadata?.deletionGracePeriodSeconds,
        resourceVersion: data.metadata?.resourceVersion,
        createdOn: parseTimePropertyValue(data.metadata?.creationTimestamp),
        deletedOn: parseTimePropertyValue(data.metadata?.deletionTimestamp),
        // spec fields
        activeDeadlineSeconds: data.spec?.activeDeadlineSeconds,
        backoffLimit: data.spec?.backoffLimit,
        completions: data.spec?.completions,
        manualSelector: data.spec?.manualSelector,
        parallelism: data.spec?.parallelism,
        ttlSecondsAfterFinished: data.spec?.ttlSecondsAfterFinished,
        'status.active': data.status?.active,
        'status.startTime': parseTimePropertyValue(data.status?.startTime),
        'status.completionTime': parseTimePropertyValue(
          data.status?.completionTime,
        ),
        'status.failed': data.status?.failed,
        'status.succeeded': data.status?.succeeded,
      },
    },
  });
}
