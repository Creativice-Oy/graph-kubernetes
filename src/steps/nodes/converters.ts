import * as k8s from '@kubernetes/client-node';
import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { Entities } from '../constants';

export function createNodeEntity(data: k8s.V1Node) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.NODE._class,
        _type: Entities.NODE._type,
        _key: data.metadata?.uid,
        name: data.metadata?.name,
        displayName: data.metadata?.name,
        generation: data.metadata?.generation,
        deletionGracePeriodSeconds: data.metadata?.deletionGracePeriodSeconds,
        resourceVersion: data.metadata?.resourceVersion,
        createdOn: parseTimePropertyValue(data.metadata?.creationTimestamp),
        // spec properties - we could group them using "spec." notation
        podCIDR: data.spec?.podCIDR,
        podCIDRs: data.spec?.podCIDRs,
        providerID: data.spec?.providerID,
        unschedulable: data.spec?.unschedulable,
        'status.allocatable.cpu':
          data.status?.allocatable && data.status?.allocatable['cpu'],
        'status.allocatable.ephemeral-storage':
          data.status?.allocatable &&
          data.status?.allocatable['ephemeral-storage'],
        'status.allocatable.hugepages-1Gi':
          data.status?.allocatable && data.status?.allocatable['hugepages-1Gi'],
        'status.allocatable.hugepages-2Mi':
          data.status?.allocatable && data.status?.allocatable['hugepages-2Mi'],
        'status.allocatable.memory':
          data.status?.allocatable && data.status?.allocatable['memory'],
        'status.allocatable.pods':
          data.status?.allocatable && data.status?.allocatable['pods'],
        'status.capacity.cpu':
          data.status?.capacity && data.status?.capacity['cpu'],
        'status.capacity.ephemeral-storage':
          data.status?.capacity && data.status?.capacity['ephemeral-storage'],
        'status.capacity.hugepages-1Gi':
          data.status?.capacity && data.status?.capacity['hugepages-1Gi'],
        'status.capacity.hugepages-2Mi':
          data.status?.capacity && data.status?.capacity['hugepages-2Mi'],
        'status.capacity.memory':
          data.status?.capacity && data.status?.capacity['memory'],
        'status.capacity.pods':
          data.status?.capacity && data.status?.capacity['pods'],
        'status.kubeletEndpointPort':
          data.status?.daemonEndpoints?.kubeletEndpoint?.Port,
        'status.nodeInfo.architecture': data.status?.nodeInfo?.architecture,
        'status.nodeInfo.bootID': data.status?.nodeInfo?.bootID,
        'status.nodeInfo.containerRuntimeVersion':
          data.status?.nodeInfo?.containerRuntimeVersion,
        'status.nodeInfo.kernelVersion': data.status?.nodeInfo?.kernelVersion,
        'status.nodeInfo.kubeProxyVersion':
          data.status?.nodeInfo?.kubeProxyVersion,
        'status.nodeInfo.kubeletVersion': data.status?.nodeInfo?.kubeletVersion,
        'status.nodeInfo.machineID': data.status?.nodeInfo?.machineID,
        'status.nodeInfo.operatingSystem':
          data.status?.nodeInfo?.operatingSystem,
        'status.nodeInfo.osImage': data.status?.nodeInfo?.osImage,
        'status.nodeInfo.systemUUID': data.status?.nodeInfo?.systemUUID,
      },
    },
  });
}
