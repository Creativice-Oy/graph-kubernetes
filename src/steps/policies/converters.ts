import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import * as k8s from '@kubernetes/client-node';
import { Entities } from '../constants';

export function createPodSecurityPolicyEntity(
  data: k8s.V1beta1PodSecurityPolicy,
) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.POD_SECURITY_POLICY._class,
        _type: Entities.POD_SECURITY_POLICY._type,
        // metadata properties
        _key: data.metadata?.uid,
        name: data.metadata?.name,
        displayName: data.metadata?.name,
        generation: data.metadata?.generation,
        deletionGracePeriodSeconds: data.metadata?.deletionGracePeriodSeconds,
        resourceVersion: data.metadata?.resourceVersion,
        createdOn: parseTimePropertyValue(data.metadata?.creationTimestamp),
        // spec
        // 5.2.5 Minimize the admission of containers with allowPrivilegeEscalation (Automated)
        allowPrivilegeEscalation: data.spec?.allowPrivilegeEscalation === true,
        // 5.2.8 Minimize the admission of containers with added capabilities (Automated)
        allowedCapabilities: data.spec?.allowedCapabilities,
        allowedProcMountTypes: data.spec?.allowedProcMountTypes,
        allowedUnsafeSysctls: data.spec?.allowedUnsafeSysctls,
        defaultAddCapabilities: data.spec?.defaultAddCapabilities,
        defaultAllowPrivilegeEscalation:
          data.spec?.defaultAllowPrivilegeEscalation,
        forbiddenSysctls: data.spec?.forbiddenSysctls,
        // 5.2.3 Minimize the admission of containers wishing to share the host IPC namespace (Automated)
        hostIPC: data.spec?.hostIPC === true,
        // 5.2.4 Minimize the admission of containers wishing to share the host network namespace (Automated)
        hostNetwork: data.spec?.hostNetwork === true,
        // 5.2.2 Minimize the admission of containers wishing to share the host process ID namespace (Automated)
        hostPID: data.spec?.hostPID === true,
        // 5.2.1 Minimize the admission of privileged containers (Automated)
        privileged: data.spec?.privileged === true,
        readOnlyRootFilesystem: data.spec?.readOnlyRootFilesystem === true,
        // 5.2.7 Minimize the admission of containers with the NET_RAW capability (Automated)
        // 5.2.9 Minimize the admission of containers with capabilities assigned (Manual)
        requiredDropCapabilities: data.spec?.requiredDropCapabilities,
        'runAsGroup.range': data.spec?.runAsGroup?.ranges?.map(
          (range) => `${range.min}, ${range.max}`,
        ),
        'runAsGroup.rule': data.spec?.runAsGroup?.rule,
        'runAsUser.range': data.spec?.runAsUser?.ranges?.map(
          (range) => `${range.min}, ${range.max}`,
        ),
        // 5.2.6 Minimize the admission of root containers (Automated)
        'runAsUser.rule': data.spec?.runAsUser?.rule,
        'runtimeClass.allowedRuntimeClassNames':
          data.spec?.runtimeClass?.allowedRuntimeClassNames,
        'runtimeClass.defaultRuntimeClassName':
          data.spec?.runtimeClass?.defaultRuntimeClassName,
        'seLinux.rule': data.spec?.seLinux.rule,
        'seLinux.seLinuxOptions.level':
          data.spec?.seLinux.seLinuxOptions?.level,
        'seLinux.seLinuxOptions.role': data.spec?.seLinux.seLinuxOptions?.role,
        'seLinux.seLinuxOptions.type': data.spec?.seLinux.seLinuxOptions?.type,
        'seLinux.seLinuxOptions.user': data.spec?.seLinux.seLinuxOptions?.user,
        'supplementalGroups.range': data.spec?.supplementalGroups.ranges?.map(
          (range) => `${range.min}, ${range.max}`,
        ),
        'supplementalGroups.rule': data.spec?.supplementalGroups?.rule,
      },
    },
  });
}

export function createNetworkPolicyEntity(data: k8s.V1NetworkPolicy) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.NETWORK_POLICY._class,
        _type: Entities.NETWORK_POLICY._type,
        // metadata properties
        _key: data.metadata?.uid,
        name: data.metadata?.name,
        displayName: data.metadata?.name,
        generation: data.metadata?.generation,
        deletionGracePeriodSeconds: data.metadata?.deletionGracePeriodSeconds,
        resourceVersion: data.metadata?.resourceVersion,
        createdOn: parseTimePropertyValue(data.metadata?.creationTimestamp),
        // spec
        // TODO: get egress data (mapping? separate entities? combination?)
        // TODO: get ingress data (mapping? separate entities? combination?)
        policyTypes: data.spec?.policyTypes,
      },
    },
  });
}
