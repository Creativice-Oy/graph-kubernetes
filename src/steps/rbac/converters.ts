import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import * as k8s from '@kubernetes/client-node';
import { Entities } from '../constants';

export function createRoleEntity(data: k8s.V1Role) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.ROLE._class,
        _type: Entities.ROLE._type,
        _key: data.metadata?.uid,
        name: data.metadata?.name,
        namespace: data.metadata?.namespace,
        displayName: data.metadata?.name,
        generation: data.metadata?.generation,
        deletionGracePeriodSeconds: data.metadata?.deletionGracePeriodSeconds,
        resourceVersion: data.metadata?.resourceVersion,
        createdOn: parseTimePropertyValue(data.metadata?.creationTimestamp),
      },
    },
  });
}

export function createClusterRoleEntity(data: k8s.V1ClusterRole) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.CLUSTER_ROLE._class,
        _type: Entities.CLUSTER_ROLE._type,
        _key: data.metadata?.uid,
        name: data.metadata?.name,
        displayName: data.metadata?.name,
        generation: data.metadata?.generation,
        deletionGracePeriodSeconds: data.metadata?.deletionGracePeriodSeconds,
        resourceVersion: data.metadata?.resourceVersion,
        createdOn: parseTimePropertyValue(data.metadata?.creationTimestamp),
      },
    },
  });
}

export function createRoleBindingEntity(data: k8s.V1RoleBinding) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.ROLE_BINDING._class,
        _type: Entities.ROLE_BINDING._type,
        _key: data.metadata?.uid,
        name: data.metadata?.name,
        displayName: data.metadata?.name,
        generation: data.metadata?.generation,
        deletionGracePeriodSeconds: data.metadata?.deletionGracePeriodSeconds,
        resourceVersion: data.metadata?.resourceVersion,
        createdOn: parseTimePropertyValue(data.metadata?.creationTimestamp),
      },
    },
  });
}

export function createClusterRoleBindingEntity(data: k8s.V1ClusterRoleBinding) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.CLUSTER_ROLE_BINDING._class,
        _type: Entities.CLUSTER_ROLE_BINDING._type,
        _key: data.metadata?.uid,
        name: data.metadata?.name,
        displayName: data.metadata?.name,
        generation: data.metadata?.generation,
        deletionGracePeriodSeconds: data.metadata?.deletionGracePeriodSeconds,
        resourceVersion: data.metadata?.resourceVersion,
        createdOn: parseTimePropertyValue(data.metadata?.creationTimestamp),
      },
    },
  });
}
