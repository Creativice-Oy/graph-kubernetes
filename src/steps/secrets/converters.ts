import * as k8s from '@kubernetes/client-node';
import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { Entities } from '../constants';

export function createSecretEntity(configMap: k8s.V1Secret) {
  // Secret contains a data which we don't want to ingest.
  const { data, ...withoutData } = configMap;

  return createIntegrationEntity({
    entityData: {
      source: withoutData,
      assign: {
        _class: Entities.SECRET._class,
        _type: Entities.SECRET._type,
        _key: withoutData.metadata!.uid!,
        name: withoutData.metadata?.name,
        displayName: withoutData.metadata?.name,
        generation: withoutData.metadata?.generation,
        deletionGracePeriodSeconds:
          withoutData.metadata?.deletionGracePeriodSeconds,
        resourceVersion: withoutData.metadata?.resourceVersion,
        createdOn: parseTimePropertyValue(
          withoutData.metadata?.creationTimestamp,
        ),
        deletedOn: parseTimePropertyValue(
          withoutData.metadata?.deletionTimestamp,
        ),
        type: withoutData.type,
        classification: null,
        encrypted: false,
      },
    },
  });
}
