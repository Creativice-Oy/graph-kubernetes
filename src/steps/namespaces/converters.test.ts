import { createNamespaceEntity } from './converters';
import { createMockNamespace } from '../../../test/mocks';

describe('#createNamespaceEntity', () => {
  test('should convert data', () => {
    expect(createNamespaceEntity(createMockNamespace())).toMatchSnapshot();
  });
});
