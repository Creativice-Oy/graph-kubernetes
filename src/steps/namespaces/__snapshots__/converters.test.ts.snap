// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createNamespaceEntity should convert data 1`] = `
Object {
  "_class": Array [
    "Group",
  ],
  "_key": "55ea52dd-382c-4b8c-8c24-bff59db4b817",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "metadata": Object {
          "creationTimestamp": 2021-04-27T20:57:17.000Z,
          "name": "example-namespace",
          "resourceVersion": "66",
          "uid": "55ea52dd-382c-4b8c-8c24-bff59db4b817",
        },
        "spec": Object {
          "finalizers": Array [
            "kubernetes",
          ],
        },
        "status": Object {
          "phase": "Active",
        },
      },
    },
  ],
  "_type": "kube_namespace",
  "active": undefined,
  "createdOn": 1619557037000,
  "deletionGracePeriodSeconds": undefined,
  "displayName": "example-namespace",
  "finalizers": Array [
    "kubernetes",
  ],
  "generation": undefined,
  "name": "example-namespace",
  "resourceVersion": "66",
  "status.phase": "Active",
}
`;
